File: lecture_1/p6.cpp
#include<iostream>
#include<vector>
using namespace std;

// O(n2)
int lis(vector<int>& arr)
{
    int strg[arr.size()];
    int ans=0;
    for(int i=0;i<arr.size();i++)
    {
        int m=0;
        for(int j=i-1;j>=0;j--)
        {
            if(arr[i]>arr[j])
                m=max(m,strg[j]);
        }
        strg[i]=m+1;
        if(strg[i]>ans)
            ans=strg[i];
    }
    return ans;
}

int main(int args,char** argv)
{
    vector<int> arr={10,3,13,7,11,12,15,18,4};
    cout<<lis(arr)<<endl;
    return 0;
}

File: lecture_24/p6.java
import java.util.*;
public class p6 {
    public static int largestRectangleArea(int[] arr) {
        Stack<Integer> st=new Stack<>();
        st.push(-1);
        int ans=0;
        for(int i=0;i<arr.length;i++)
        {
            while(st.peek()!=-1 && arr[i]<arr[st.peek()])
            {
                int h=arr[st.pop()];
                int w=i-st.peek()-1;
                int a=h*w;
                ans=Math.max(ans,a);
            }
            st.push(i);
        }
        while(st.peek()!=-1)
        {
            int h=arr[st.pop()];
            int w=arr.length-st.peek()-1;
            int a=h*w;
            ans=Math.max(ans,a);
        }
        return ans;
    }
    public static void main(String[] args) {
        int arr[]={2,1,5,6,2,3};
        System.out.println(largestRectangleArea(arr));
    }
}